File: min0_forth.rb
# frozen_string_literal: true

require_relative "min0_core_forth_dictionary"
require_relative "min0_core_forth_outer"

MIN0_FORTH_VERSION = File.read(File.join(__dir__, "VERSION"), encoding: "UTF-8").strip
MIN0_FORTH_BANNER = "MIN0 CORE FORTH #{MIN0_FORTH_VERSION} - educational and experimental reference"

def make_min0_host
  vm = Min0CoreForth::VM.new
  dictionary = Min0CoreForth::RuntimeDictionary.new(vm)
  Min0CoreForth.install_core_primitives(dictionary)
  Min0CoreForth::OuterInterpreter.new(vm, dictionary)
end

def write_new_min0_output(outer, start)
  fragments = outer.output[start..] || []
  unless fragments.empty?
    $stdout.write(fragments.join)
    $stdout.flush
  end
  outer.output.length
end

def run_min0_file(path, quiet:)
  outer = make_min0_host
  begin
    outer.interpret(File.read(path, encoding: "UTF-8"))
  rescue StandardError => e
    warn "ERROR #{e.class.name.split('::').last}: #{e.message}"
    return 1
  end
  puts MIN0_FORTH_BANNER unless quiet
  write_new_min0_output(outer, 0)
  unless quiet
    puts if !outer.output.empty? && !outer.terminal_text.end_with?("\n")
    puts "DATA stack: #{outer.vm.data_stack.inspect}"
  end
  0
end

def run_min0_repl
  outer = make_min0_host
  output_index = 0
  puts MIN0_FORTH_BANNER
  puts "Type BYE or EXIT to leave. Output bytes are emitted by the program."
  loop do
    $stdout.write("ok> ")
    $stdout.flush
    source = $stdin.gets
    if source.nil?
      puts
      return 0
    end
    return 0 if %w[BYE EXIT].include?(source.strip.upcase)

    begin
      outer.interpret(source)
      output_index = write_new_min0_output(outer, output_index)
      puts if !outer.output.empty? && !outer.terminal_text.end_with?("\n")
      puts " ok #{outer.vm.data_stack.inspect}"
    rescue StandardError => e
      warn "ERROR #{e.class.name.split('::').last}: #{e.message}"
    end
  end
end

def min0_usage
  "usage: ruby min0_forth.rb [SOURCE] | -z SOURCE | --version"
end

exit_code = case ARGV
            in ["--version"]
              puts MIN0_FORTH_VERSION
              0
            in ["-z" | "--quiet-source", path]
              run_min0_file(path, quiet: true)
            in [path]
              run_min0_file(path, quiet: false)
            in []
              run_min0_repl
            else
              warn min0_usage
              2
            end
exit exit_code

File: min0_forth.py
"""User-facing host launcher for MIN0 CORE FORTH."""

from __future__ import annotations

import argparse
import sys
from pathlib import Path

from min0_core_forth_dictionary import RuntimeDictionary
from min0_core_forth_outer import OuterInterpreter, install_core_primitives
from min0_core_forth_vm import Min0CoreForthVM


ROOT = Path(__file__).resolve().parent
VERSION = (ROOT / "VERSION").read_text(encoding="utf-8").strip()
BANNER = f"MIN0 CORE FORTH {VERSION} - educational and experimental reference"


def make_host() -> OuterInterpreter:
    vm = Min0CoreForthVM()
    dictionary = RuntimeDictionary(vm)
    install_core_primitives(dictionary)
    return OuterInterpreter(vm, dictionary)


def execute_source(outer: OuterInterpreter, source: str) -> None:
    outer.interpret(source)


def _write_new_output(outer: OuterInterpreter, start: int) -> int:
    fragments = outer.output[start:]
    if fragments:
        sys.stdout.write("".join(fragments))
        sys.stdout.flush()
    return len(outer.output)


def run_file(path: Path, *, quiet: bool) -> int:
    outer = make_host()
    try:
        source = path.read_text(encoding="utf-8")
        execute_source(outer, source)
    except Exception as exc:
        print(f"ERROR {type(exc).__name__}: {exc}", file=sys.stderr)
        return 1
    if not quiet:
        print(BANNER)
    _write_new_output(outer, 0)
    if not quiet:
        if outer.output and not outer.terminal_text.endswith("\n"):
            print()
        print(f"DATA stack: {outer.vm.data_stack}")
    return 0


def repl() -> int:
    outer = make_host()
    output_index = 0
    print(BANNER)
    print("Type BYE or EXIT to leave. Output bytes are emitted by the program.")
    while True:
        try:
            source = input("ok> ")
        except EOFError:
            print()
            return 0
        if source.strip().upper() in {"BYE", "EXIT"}:
            return 0
        try:
            execute_source(outer, source)
            output_index = _write_new_output(outer, output_index)
            if outer.output and not outer.terminal_text.endswith("\n"):
                print()
            print(f" ok {outer.vm.data_stack}")
        except Exception as exc:
            print(f"ERROR {type(exc).__name__}: {exc}", file=sys.stderr)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(
        description="Run the MIN0 CORE FORTH Python host reference"
    )
    parser.add_argument("source", nargs="?", type=Path, help="FORTH source file")
    parser.add_argument(
        "-z", "--quiet-source", metavar="FILE", type=Path,
        help="run FILE without banner, prompt, or final stack",
    )
    parser.add_argument("--version", action="store_true")
    args = parser.parse_args(argv)
    if args.version:
        print(VERSION)
        return 0
    if args.source is not None and args.quiet_source is not None:
        parser.error("source and -z cannot be used together")
    if args.quiet_source is not None:
        return run_file(args.quiet_source, quiet=True)
    if args.source is not None:
        return run_file(args.source, quiet=False)
    return repl()


if __name__ == "__main__":
    raise SystemExit(main())
